### PR TITLE
[MM-37629] Include non-members and auto-invite users to owner/assignee dropdowns.

### PR DESCRIPTION
**Pull request by shadowshot-x**
#### Summary
Adds Non-Channel Members to the Channel as Owners. Auto-Invite and add the non channel members.

#### Ticket Link

Fixes: https://github.com/mattermost/mattermost-server/issues/18590

Screenshot from 2021-12-08 22-13-09
Screenshot from 2021-12-08 22-13-19
Screenshot from 2021-12-08 22-16-52

#### Checklist

- ~[ ] Modify the Modal to add Non-Channel Members~
- ~[ ] Add Modal on non-channel members click~
- ~[ ] Add the added member as owner~

### CODE REVIEW

**Comment by agarciamontoro** [comment]
Thank you for this, @shadowshot-x! Christopher and I will review this as soon as we can :)

Just an early FYI before the review: it seems like there's a conflict in the `rhs_about.tsx` file, could you fix that? Thank you :pray:

**Comment by shadowshot-x** [comment]
Hi @agarciamontoro. Thanks for the heads up. I just saw and fixed it. PS. Kudos to the team as I can see you guys brought the E2E tests time down! :100:

**Comment by agarciamontoro** [comment]
> Kudos to the team as I can see you guys brought the E2E tests time down! 100

Kudos to @Phrynobatrachus in particular!

I'll have to review this on Monday, didn't have time today :( Sorry for the inconvenience, @shadowshot-x!

**Comment by shadowshot-x** [comment]
No issues @agarciamontoro

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
This is amazing, @shadowshot-x, thank you for all your work! I tested it and worked smoothly, but I left a bunch of comments to make the code a bit simpler.

File: webapp/src/components/backstage/runs_list/filters.tsx [hunk 1]
```diff
@@ -178,6 +178,7 @@ const Filters = ({fetchParams, setFetchParams, fixedTeam, fixedPlaybook}: Props)
                 }}
                 controlledOpenToggle={profileSelectorToggle}
                 getUsers={fetchOwners}
+                getUsersInTeam={fetchOwners}
```
[the remark this answers is not in the document]
Comment (shadowshot-x) [reply webapp/src/components/backstage/runs_list/filters.tsx:181]: Good catch! I will update it.

File: webapp/src/components/profile/profile_selector.tsx [hunk 2]
```diff
@@ -21,6 +21,8 @@ export interface Option {
     value: string;
     label: JSX.Element | string;
     userId: string;
+    userType: string;
+    userObj: UserProfile;
```
[the remark this answers is not in the document]
Comment (shadowshot-x) [reply webapp/src/components/profile/profile_selector.tsx:25]: Yes we would still need `userType`. I will eliminate `userId` from here.

File: webapp/src/components/profile/profile_selector.tsx [hunk 3]
```diff
@@ -94,6 +98,24 @@ export default function ProfileSelector(props: Props) {
         };
 
         const users = await props.getUsers();
+        const usersInTeam = await props.getUsersInTeam();
```
[the remark this answers is not in the document]
Comment (shadowshot-x) [reply webapp/src/components/profile/profile_selector.tsx:101]: Agreed.

File: webapp/src/components/rhs/rhs_about.tsx [hunk 4]
```diff
@@ -31,14 +41,38 @@ const RHSAbout = (props: Props) => {
...
+
+    const team = useSelector<GlobalState, Team>((state) => getTeam(state, channel?.team_id));
+    const channelNamesMap = useSelector<GlobalState, ChannelNamesMap>(getChannelsNameMapInCurrentTeam);
+    const [showAddToChannel, setShowAddToChannelConfirm] = useState(false);
+    const [currentUserSelect, setCurrentUserSelect] = useState<UserProfile | null>();
+    const teamnameNameDisplaySetting = useSelector<GlobalState, string | undefined>(getTeammateNameDisplaySetting) || '';
+    const overviewURL = `/playbooks/runs/${props.playbookRun.id}`;
+
+    const markdownOptions = {
+        singleline: false,
```
[the remark this answers is not in the document]
Comment (shadowshot-x) [reply webapp/src/components/rhs/rhs_about.tsx:56]: Done, will push in next commit.

File: webapp/src/actions.ts [hunk 5]
```diff
@@ -131,6 +133,14 @@ export function addNewTask(checklist: number) {
...
 }
 
+export function addToChannel(userId: string) {
+    return async (dispatch: Dispatch<AnyAction>, getState: GetStateFunc) => {
+        const currentChannelId = getCurrentChannelId(getState());
+
+        await clientAddChannelMember(dispatch, getState, currentChannelId, userId);
+    };
+}
+
```
[the remark this answers is not in the document]
Comment (shadowshot-x) [reply webapp/src/actions.ts:143]: Sure, this is an easier way. I was following the functions in actions.ts and so I took an await based approach. Will change this in the next commit.

File: webapp/src/client.ts [hunk 6]
```diff
@@ -156,6 +156,33 @@ export function fetchPlaybookRunChannels(teamID: string, userID: string) {
...
+        {
+            type: ChannelTypes.RECEIVED_CHANNEL_MEMBER,
+            data: member,
+        },
+        {
+            type: ChannelTypes.ADD_CHANNEL_MEMBER_SUCCESS,
+            id: channelId,
+        },
+    ], 'ADD_CHANNEL_MEMBER.BATCH'));
+}
```
[the remark this answers is not in the document]
Comment (shadowshot-x) [reply webapp/src/client.ts:184]: Right.

File: webapp/src/components/profile/profile_selector.tsx [hunk 7]
```diff
@@ -255,7 +279,8 @@ export default function ProfileSelector(props: Props) {
                 hideSelectedOptions={false}
                 isClearable={props.isClearable}
                 menuIsOpen={true}
-                options={userOptions}
+                options={[{label: 'CHANNEL MEMBERS', options: userOptions},
+                    {label: 'NOT IN CHANNEL', options: userNotInChannelOptions}]}
```
[the remark this answers is not in the document]
Comment (shadowshot-x) [reply webapp/src/components/profile/profile_selector.tsx:283]: True.

File: webapp/src/components/rhs/rhs_about.tsx [hunk 8]
```diff
@@ -31,14 +41,38 @@ const RHSAbout = (props: Props) => {
     const channelId = useSelector(getCurrentChannelId);
     const profilesInChannel = useProfilesInCurrentChannel();
     const collapsed = useSelector(currentRHSAboutCollapsedState);
+    const channelName = useSelector(getCurrentChannelNameForSearchShortcut);
+    const channel = useSelector<GlobalState, Channel>((state) => getChannel(state, channelId));
```
[the remark this answers is not in the document]
Comment (shadowshot-x) [reply webapp/src/components/rhs/rhs_about.tsx:45]: Right, this simplifies the import. I will change this.

File: webapp/src/client.ts [hunk 9]
```diff
@@ -4,10 +4,10 @@
 import {AnyAction, Dispatch} from 'redux';
 import qs from 'qs';
 
-import {GetStateFunc} from 'mattermost-redux/types/actions';
+import {GetStateFunc, batchActions} from 'mattermost-redux/types/actions';
```
Comment (agarciamontoro) [inline webapp/src/client.ts:7]: We can probably remove this new import, right?
```suggestion
import {GetStateFunc} from 'mattermost-redux/types/actions';
```

File: webapp/src/client.ts [hunk 10]
```diff
@@ -4,10 +4,10 @@
...
 import {UserProfile} from 'mattermost-redux/types/users';
 import {Channel} from 'mattermost-redux/types/channels';
-import {IntegrationTypes} from 'mattermost-redux/action_types';
+import {IntegrationTypes, UserTypes, ChannelTypes} from 'mattermost-redux/action_types';
```
Comment (agarciamontoro) [inline webapp/src/client.ts:10]: And this one!
```suggestion
import {IntegrationTypes} from 'mattermost-redux/action_types';
```

File: webapp/package-lock.json [hunk 11]
```diff
@@ -35,7 +35,7 @@
         "@babel/preset-react": "7.14.5",
         "@babel/preset-typescript": "7.14.5",
         "@formatjs/cli": "4.3.1",
-        "@pmmmwh/react-refresh-webpack-plugin": "^0.4.3",
```
Comment (agarciamontoro) [inline webapp/package-lock.json:38]: It looks like your last commit added modifications to this file by accident. Could you remove them?

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Thank you, @shadowshot-x, it looks amazing! :tada:

I added a couple of comments below, but also when testing I found a minor issue (sorry for not catching it before):
As the `ProfileSelector` component now adds labels to the options, the filter in the runs list also shows `CHANNEL MEMBERS` (see screenshot below). It does not make sense to add the label in that case (there's no channel we can refer to), so what do you think about adding the labels only when the `userNotInChannelOptions` is not empty? If we only have elements in the `userOptions` array, it should show no labels at all.
image

File: webapp/package-lock.json (hunk 11 above)
Comment (shadowshot-x) [reply webapp/package-lock.json:38]: Sure, I will revert it back.

File: webapp/src/client.ts (hunk 9 above)
Comment (shadowshot-x) [reply webapp/src/client.ts:7]: Sure, will do in the next commit.

File: webapp/src/client.ts (hunk 10 above)
Comment (shadowshot-x) [reply webapp/src/client.ts:10]: Yes.

**Comment by shadowshot-x** [comment]
@agarciamontoro, I totally agree. It does not make sense to show labels in that case. I will update the code for this.

**Review by cpoile — APPROVED** [review]
This is great work @shadowshot-x, very nicely done (and excellent work addressing @agarciamontoro 's comments). Thank you!

**Comment by shadowshot-x** [comment]
Thank you @agarciamontoro for testing the PR. I also tested a bit on my end. I saw that I was breaking the `filters.tsx` ie. Filters were not being applied based on the owner. I have corrected it now. Also, I did the changes you asked for.

**Review by agarciamontoro — APPROVED** [review]
Incredible work, @shadowshot-x! Thank you again for all your efforts and patience :tada:

**Pull request merged**